homeassistant/util/__init__.py

```diff
@@ -286,6 +286,62 @@ def wrapper(*args, **kwargs):
         return wrapper
 
 
+class Synchronized(object):
+    """
+    A method decorator to add locking to a method to prevent it from being
+    called in parallel.
+
+    Decorator takes in one 'shared_lock' parameter that creates a lock that is
+    shared between all other decorated functions with the parameter set.
+
+    """
+    # pylint: disable=too-few-public-methods
+
+    decorator_lock = threading.Lock()
+
+    def __init__(self, shared_lock=False):
+        self.shared_lock = shared_lock
+
+    def __call__(self, method):
+        # Different methods that can be passed in:
+        #  - a function
+        #  - an unbound function on a class
+        #  - a method (bound function on a class)
+
+        # We want to be able to differentiate between function and unbound
+        # methods (which are considered functions).
+        # All methods have the classname in their qualname seperated by a '.'
+        # Functions have a '.' in their qualname if defined inline, but will
+        # be prefixed by '.<locals>.' so we strip that out.
+        is_func = (not hasattr(method, '__self__') and
```

Comment:
There is a lot of duplicate logic that could be extracted between Throttle/Synchronized. Like this method or determining what the host is.